File: medium/0055. Jump Game/main.go
// Solved by @kitanoyoru

package main

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func canJump(nums []int) bool {
	maxJump := 0
	for i := 0; i <= maxJump; i++ {
		maxJump = max(maxJump, i+nums[i])
		if maxJump >= len(nums)-1 {
			return true
		}
	}

	return false
}

File: medium/0055. Jump Game/main.cpp
// Solved by @kitanoyoru
// https://leetcode.com/problems/jump-game/

#include <vector>
#include <cmath>

using namespace std;

class Solution {
public:
  bool canJump(vector<int>& nums) {
    int max_jump = 0;
    for (int i = 0; i <= max_jump; i++) {
      max_jump = max(max_jump, i + nums[i]);
      if (max_jump >= nums.size() - 1) {
        return true;
      }
    }

    return false;
  }
};
